src/FSharp.Core/array.fs

```diff
@@ -1588,6 +1587,32 @@ module Array =
 
         acc
 
+    let isNetFramework = System.Runtime.InteropServices.RuntimeInformation.FrameworkDescription.StartsWith ".NET Framework"
```

Comment:
Is this going to functionally compile to a `static readonly bool`, so that the JIT can optimize things appropriately?

Is this going to do the "wrong" thing on custom runtimes or scenarios where vectorization may not be available or possible? For example, there was no SIMD acceleration on 32-bit Unix for a while and there is non on Arm32 today. Likewise, acceleration can be disabled via environment variables for testing purposes.

In general it's expected that `Enumerable.Sum` is going to do the most optimal thing over time based on the underlying hardware and other user options (like if you're compiling for size vs speed).

Reply:
> Is this going to functionally compile to a static readonly bool, so that the JIT can optimize things appropriately?

This is runtime check, not compile-time check. Because nothing says the code is compiled and run on similar machines.

> Is this going to do the "wrong" thing on custom runtimes or scenarios where vectorization may not be available or possible? For example, there was no SIMD acceleration on 32-bit Unix for a while and there is non on Arm32 today. Likewise, acceleration can be disabled via environment variables for testing purposes.

No, because Enumerable.Sum already checks that within its implementation.
The only reason for this check is because Enumerable.Sum is slow on old .NET Framework.

Reply:
> This is runtime check, not compile-time check. Because nothing says the code is compiled and run on similar machines.

Yes, which is why it should compile down to a `static readonly bool` in IL. Because that will cause it to be initialized at runtime in Tier 0 and then allow the JIT to treat it as a constant in Tier 1 (or for NativeAOT), allowing the check to be elided once we do know the actual machine/runtime it's running on.

> The only reason for this check is because Enumerable.Sum is slow on old .NET Framework.

The point was that you're doing a specific check for .NET Framework, which doesn't account for custom runtimes or other scenarios that may or may not be relevant. So I'm just asking if the nuance of that has been fully considered.

Reply:
@T-Gro I think he means that the check should be optimized away by JIT and directly use the relevant implementation depending on framework - `Array.sum` on .NET Framework, `Enumerable.Sum` otherwise.